src/abstractdataframe/selection.jl

```diff
@@ -170,56 +169,207 @@ function normalize_selection(idx::AbstractIndex,
     return (wanttable ? AsTable(c) : c) => fun => newcol
 end
 
-function select_transform!(nc::Pair{<:Union{Int, AbstractVector{Int}, AsTable},
-                                    <:Pair{<:Base.Callable, Symbol}},
+function _transformation_helper(df, col_idx, @nospecialize(fun))
+    if col_idx === nothing
+        return fun(df)
+    elseif col_idx isa Int
+        return fun(df[!, col_idx])
+    elseif col_idx isa AsTable
+        tbl = Tables.columntable(select(df, col_idx.cols, copycols=false))
+        if isempty(tbl) && fun isa ByRow
+            if isempty(df)
+                T = Base.return_types(fun.fun, (NamedTuple{(),Tuple{}},))[1]
+                return T[]
+            else
+                return [fun.fun(NamedTuple()) for _ in 1:nrow(df)]
+            end
```

Comment:
> since `collect` will use inference to choose the element type of the empty vector?

It was failing for some reason - I will recheck and remove `if` it it is possible